flink-metrics/flink-metrics-statsd/src/main/java/org/apache/flink/metrics/statsd/StatsDReporter.java

```diff
@@ -179,41 +254,130 @@ private String prefix(String ... names) {
 		}
 	}
 
-	private void send(final String name, final String value) {
+	private String buildStatsdLine(final String name, final String value, final String tags) {
+		Double number;
 		try {
-			String formatted = String.format("%s:%s|g", name, value);
-			byte[] data = formatted.getBytes(StandardCharsets.UTF_8);
-			socket.send(new DatagramPacket(data, data.length, this.address));
+			number = Double.parseDouble(value);
+		}
+		catch (NumberFormatException e) {
+			// quietly skip values like "n/a"
+			return "";
 		}
-		catch (IOException e) {
-			LOG.error("unable to send packet to statsd at '{}:{}'", address.getHostName(), address.getPort());
+		if (number >= 0.) {
+			return String.format("%s:%s|g%s", name, value, tags != null ? tags : "");
+		} else {
+			// quietly skip "unknowns" like lowWaterMark:-9223372036854775808, or JVM.Memory.NonHeap.Max:-1, or NaN
+			return "";
 		}
 	}
 
-	@Override
-	public String filterCharacters(String input) {
+	private void send(final String name, final String value, final String tags) {
+		String formatted = buildStatsdLine(name, value, tags);
+		if (formatted.length() > 0) {
+			try {
+				byte[] data = formatted.getBytes(StandardCharsets.UTF_8);
+				socket.send(new DatagramPacket(data, data.length, this.address));
+			}
+			catch (IOException e) {
+				LOG.error("unable to send packet to statsd at '{}:{}'", address.getHostName(), address.getPort());
+			}
+		}
+	}
+
+	/**
+	* dogstatsd names should: start with letter, uses ascii alphanumerics and underscore, separated by periods.
+	* Collapse runs of invalid characters into an underscore. Discard invalid prefix and suffix.
+	* Eg: ":::metric:::name:::" ->  "metric_name"
+	*/
+
+	private boolean isValidStatsdChar(char c) {
+		return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || (c == '_');
```

Comment:
From what I have seen '-' seems to be ok at least for some implementations. There may be other safe characters too. In fact the ReadyTalk implementation replaces spaces with '-'.
https://github.com/ReadyTalk/metrics-statsd/blob/master/metrics-statsd-common/src/main/java/com/readytalk/metrics/StatsD.java#L129

I'm sure many other characters are ok too. The main ones that need to be escaped are ones that have special meaning in the statsd format.